[BPF/SNAT] Improve code readability and complexity of the datapath

sahid opened the pull request:
This is addressing issue  #19660.

> Split the snat_v{4,6}_process() functions into nat_snat_v{4,6}() and nat_rev_snat_v{4,6}().
> 
> The proposed change would not only improve the readability of the datapath code, but also should reduce the verifier complexity of the code path which is responsible for doing the rev-SNAT translations, as no new SNAT mapping needs to be find for this code path (only the lookup in the BPF SNAT map).

Fixes: #19660 

```release-note
Masquerading bpf mode -  Improve code readability and comlexity of the datapath.
```

sahid committed:
bpf/nat: introduce snat_v4_nat() and snat_v4_rev_nat functions

This split snat_v4_process() to introduce two functions: snat_v4_nat()
and snat_v4_rev_nat().  This does not change logic, it is expected to
reduce verifier complexity and improve code maintenability.

Fixes: #19660
Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

brb reviewed (changes requested):
I'd suggest to split this PR into multiple:

1. Introduce the new tail call.
2. Split `snat_v4_process()` into `nat_do_snat()` and `nat_do_rev_snat()`.

Another thing, we recently introduced the eBPF unit testing framework, so your tests needs to be moved to it. Please see https://docs.cilium.io/en/latest/contributing/testing/bpf/.

sahid committed:
bpf/nat: introduce snat_v6_nat() and snat_v6_rev_nat functions

This split snat_v6_process() to introduce two functions: snat_v6_nat()
and snat_v6_rev_nat().  This does not change logic, it is expected to
reduce verifier complexity and improve code maintenability.

Fixes: #19660

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

brb commented on bpf/lib/nat.h:
```diff
@@ -558,22 +558,95 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+		return NAT_PUNT_TO_STACK;
+	};
+
+	if (snat_v4_can_skip(target, &tuple, NAT_DIR_INGRESS, from_endpoint, icmp_echoreply))
```
I haven't done extensive review, but sharing my thoughts - I think we can take one step further with this refactoring. I suggest to split `snat_v*_can_skip` into two functions - one for SNAT, one for rev-SNAT. And then the same for `snat_v*_handle_mapping`, so that the one responsible for rev-SNAT would do only the lookup w/o trying to find a free mapping.

sahid replied:
Sure, makes sense, thank you.

sahid replied:
@brb a gentle ping, I have updated the PR with your requested changes. Not sure that you get a notif for that.

sahid committed:
bpf/nat: introduce snat_v*_can_skip() nat and rev_nat

In this desire of continuing improving code quality and verifier
complexity. This patch is splitting snat_v*_can_skip into two
functions.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

sahid committed:
bpf/nat: introduce snat_v*_handle_mapping() nat and rev_nat

In this desire of continuing improving code quality and verifier
complexity. This patch is splitting snat_v*_handle_mapping into two
functions.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

qmonnet commented on bpf/lib/nat.h:
```diff
@@ -558,22 +558,95 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+			break;
+		default:
+			return DROP_NAT_UNSUPP_PROTO;
+		}
```
I understand the desire for simplification with regards to the `if`s in `snat_v4_nat()` here, but I'd rather keep the same code between the two functions if there's no reason to change. Could you please either revert to `if`s, or update `snat_v4_nat()` to use a `switch/case` as well?

sahid replied:
oh right,  I should have not touch this. That part is coming from an other work that I have in progress #18414 for some reasons I have made some mix.

qmonnet commented on bpf/lib/nat.h:
```diff
@@ -570,8 +571,7 @@ snat_v4_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target,
 static __always_inline __maybe_unused int
-snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target,
-		bool from_endpoint)
+snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
```
We have two versions of `snat_v4_rev_nat()`, but you forgot to update the arguments for the other one (in the other branch of `#if defined(ENABLE_IPV4) && defined(ENABLE_NODEPORT)`). Please update it, too

qmonnet commented on bpf/lib/nat.h:
```diff
@@ -626,7 +626,7 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target,
 	};
 
-	if (snat_v4_can_skip(target, &tuple, NAT_DIR_INGRESS, from_endpoint, icmp_echoreply))
+	if (snat_v4_rev_nat_can_skip(target, &tuple))
```
I think `icmp_echoreply` is no longer necessary in `snat_v4_rev_nat()`, and can be removed?

qmonnet commented on bpf/lib/nat.h:
```diff
@@ -1186,7 +1187,7 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
 	};
 
-	if (snat_v6_can_skip(target, &tuple, NAT_DIR_INGRESS, icmp_echoreply))
+	if (snat_v6_rev_nat_can_skip(target, &tuple))
```
Same remark, `icmp_echoreply` no longer in use.

qmonnet commented on bpf/lib/nat.h:
```diff
@@ -298,27 +298,43 @@ static __always_inline int snat_v4_track_local(struct __ctx_buff *ctx,
-						  struct ipv4_nat_entry *tmp,
-						  enum nat_dir dir, __u32 off,
-						  const struct ipv4_nat_target *target)
+static __always_inline int snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
```
Nit: It's a long line, I'd return after the return type.

```suggestion
static __always_inline int
snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
```

qmonnet reviewed (changes requested):
Nice work, thanks! Please fine some comments inline below.

qmonnet reviewed (approved):
Looks good to me, thanks for this work!

brb commented on bpf/lib/nat.h:
```diff
@@ -558,24 +577,90 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+	tuple.daddr = ip4->daddr;
+	tuple.saddr = ip4->saddr;
+	tuple.flags = NAT_DIR_INGRESS;
+	off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
```
Could we move this to an inlined helper function to reduce the boilerplate?

sahid replied:
Hum.. can you be more specific? It seems that may be related to a larger refactoring as not only nat.h is using this pattern. I can open an other PR to make that clean if you are agree.

brb replied:
I was suggesting to add a helper function to fill the tuple, as the tuple extraction code in `snat_v4_nat()` and `snat_v4_rev_nat()` looks the same (same for the v6).

sahid replied:
Honestly even aligned I'm a bit mitigated doing this change, I understand that you want to avoid any redundant part but isn't that too much at that point ? :-) Please let me know, if you really want it I will make this to happen like a charm.

pchaigno replied:
I agree with Sahid here. It won't be trivial to share the code between IPv4 and IPv6 and we already have such code inlined in other places (e.g., bpf_lxc).

brb replied:
I meant two helper functions - one for ipv4, one for ipv6. For instance, both `snat_v4_nat()` and `snat_v4_rev_nat()` share the same tuple extraction code. Any duplicate code is a potential source for bugs.

sahid replied:
Sure, done.

sahid committed:
bpf/nat: introduce snat_v*_init_tuple() to avoid duplicate code

This is introducting and using snat_v*_init_tuple() to set initial
values for the ct tuple.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

sahid committed:
bpf/nat: remove unnecessary redondante nexthdr variable

It's redondante variable that does not look to bring any value.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

brb commented on bpf/lib/nat.h:
```diff
@@ -515,9 +534,19 @@ static __always_inline __maybe_unused int snat_v4_create_dsr(struct __ctx_buff *
+	tuple->nexthdr = ip4->protocol;
+	tuple->daddr = ip4->daddr;
+	tuple->saddr = ip4->saddr;
+	tuple->flags = dir;
```
Thanks for the change. However, I mean to move the whole tuple extraction into the helper - not only L3, but also L4 (port extraction / ICMP ID is still outside the helper).

sahid replied:
Hum... I would have appreciated more details in the previous set of comments ;-)  As mentioned I was mitigated doing this change and now understanding that you was also requesting L4 changes but I don't think that will work.

We are not expecting nat and rev-nat to share the same kind of features. For instance, the support of ICMP DestUnreach need to frag only makes sense for rev-nat.

brb replied:
> For instance, the support of ICMP DestUnreach need to frag only makes sense for rev-nat.

Oh, I see. Sorry, I was not aware that the dst unreach will be needed only for rev-NAT.

brb commented on bpf/lib/nat.h:
```diff
@@ -572,24 +599,88 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+	off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
+	switch (tuple.nexthdr) {
+	case IPPROTO_TCP:
+	case IPPROTO_UDP:
```
I thin you forgot to add:
```
#ifdef ENABLE_SCTP
	case IPPROTO_SCTP:
#endif /* ENABLE_SCTP */
```

sahid replied:
Oh right, thank you.
